### PR TITLE
[9.0][MIG] medical

### PR DESCRIPTION
**Pull request by lasley**
I also changed the License to markdown version - looks nicer IMO

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I'm not sure if the license file should be changed as is a verbatim copy of the one provided by the FSF...

**Comment by lasley** [comment]
@pedrobaeza No problem, updated

File: medical/README.rst [hunk 1]
```diff
@@ -71,7 +71,11 @@ Bug Tracker
 Bugs are tracked on `GitHub Issues <https://github.com/OCA/vertical-medical/issues>`_.
 In case of trouble, please check there if your issue has already been reported.
 If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
-`here <https://github.com/OCA/vertical-medical/issues/new?body=module:%20medical%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+<<<<<<< HEAD
+`here <https://github.com/OCA/vertical-medical/issues/new?body=module:%20medical%0Aversion:%209.0.1.1.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+=======
+`here <https://github.com/OCA/vertical-medical/issues/new?body=module:%20medical%0Aversion:%209.0.2.0.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+>>>>>>> feature/SMD-88-upgrade-allergy-support-in-medical.-9.0
```
Comment (JayVora-SerpentCS) [inline medical/README.rst]: Seems conflicted.

File: medical/README.rst [hunk 2]
```diff
@@ -71,7 +71,7 @@ Bug Tracker
 Bugs are tracked on `GitHub Issues <https://github.com/OCA/vertical-medical/issues>`_.
 In case of trouble, please check there if your issue has already been reported.
 If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
-`here <https://github.com/OCA/vertical-medical/issues/new?body=module:%20medical%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+`here <https://github.com/OCA/vertical-medical/issues/new?body=module:%20medical%0Aversion:%209.0.2.0.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (pedrobaeza) [inline medical/README.rst]: Please put here only branch (9.0)

File: medical/__openerp__.py [hunk 3]
```diff
@@ -1,11 +1,14 @@
 # -*- coding: utf-8 -*-
 ##############################################################################
+#
 #    OpenERP, Open Source Management Solution
```
Comment (pedrobaeza) [inline medical/__openerp__.py]: Please put short headers like https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py

File: medical/__openerp__.py [hunk 4]
```diff
@@ -37,7 +42,7 @@
 ##############################################################################
 {
     'name': 'Odoo Medical',
-    'version': '8.0.1.1.0',
+    'version': '9.0.2.0.0',
```
Comment (pedrobaeza) [inline medical/__openerp__.py]: You can start in 9.0.1.0.0, as it's greater than 8.0.1.1.0

File: medical/models/medical_patient.py [hunk 5]
```diff
@@ -33,73 +32,90 @@ class MedicalPatient(models.Model):
...
     _inherits = {'res.partner': 'partner_id', }
 
-    @api.one
-    def _compute_age(self):
+    age = fields.Char(
+        compute='_compute_age',
+    )
+    identification_code = fields.Char(
+        string='Internal Identification',
+        help=_(
```
Comment (pedrobaeza) [inline medical/models/medical_patient.py]: You don't have to call the translate method `_` here. Translations are already handled by Odoo without it.

File: medical/models/medical_patient.py [hunk 6]
```diff
@@ -33,73 +32,90 @@ class MedicalPatient(models.Model):
...
-        else:
-            years_months_days = _('No DoB !')
-        self.age = years_months_days
+                years_months_days = _('No DoB !')
+            rec_id.age = years_months_days
 
-    @api.one
-    def action_invalidate(self):
+    @api.multi
+    def action_invalidate(self, ):
```
Comment (pedrobaeza) [inline medical/models/medical_patient.py]: If you put `api.multi`, then you have to iterate over self for getting each possible record. You don't need also the final comma

File: medical/models/medical_patient.py [hunk 7]
```diff
@@ -33,73 +32,90 @@ class MedicalPatient(models.Model):
...
-            ('w', 'Widowed'),
-            ('d', 'Divorced'),
-            ('x', 'Separated'),
-            ('z', 'law marriage'),
-        ])
-
     @api.model
     @api.returns('self', lambda value: value.id)
-    def create(self, vals):
+    def create(self, vals, ):
```
Comment (pedrobaeza) [inline medical/models/medical_patient.py]: Why adding the final comma?

File: medical/models/medical_physician.py [hunk 8]
```diff
@@ -98,10 +98,3 @@ class MedicalPhysician(models.Model):
     )
 
     _defaults = {'is_doctor': True, 'supplier': True, 'active': True, }
-
-    @api.model
-    def create(self, vals,):
-        groups_proxy = self.env['res.groups']
-        group_ids = groups_proxy.search([('name', '=', 'Medical Doctor')])
-        vals['groups_id'] = [(6, 0, group_ids)]
```
Comment (pedrobaeza) [inline medical/models/medical_physician.py:106]: Please put this as a default value, and use `self.env.ref` for searching through an XML-ID.

**Comment by pedrobaeza** [comment]
I have made a rough code review. You're removing entirely some code and views, is this moved to another module?

**Comment by lasley** [comment]
Thanks @pedrobaeza - 

Will apply the recommended fixes (and will apply recommendations to the other PRs where applicable).

All of the `medical_appointment` logic was moved to two separate modules (sorry I should have noted this in the comments of the PR):
- `medical_base_history` - Contains the a slightly upgraded/more abstract form of the change tracking used in `medical_appointment`
- `medical_appointment` - The actual appointment logic

Both of these are in #70

File: medical/models/medical_patient.py (hunk 7 above)
Comment (lasley) [inline medical/models/medical_patient.py]: Just habit as it tends to be convenient later on; but mainly with arrays/dicts, not methods. Will change

File: medical/__openerp__.py (hunk 4 above)
Comment (lasley) [inline medical/__openerp__.py]: @pedrobaeza - Just so I'm 100% - the upgrade from Odoo is already considered breaking & in requirement of migrations, hence the downgrade to 1.0.0 internal versioning?

File: medical/models/medical_physician.py (hunk 8 above)
Comment (lasley) [inline medical/models/medical_physician.py:106]: Mind if we push this off? 

We're in the process of breaking `medical_physician` into it's own module, and this change has already been made there - https://github.com/laslabs/vertical-medical/tree/feature/SMD-115-upgrade-medical_physician_to_9.0/medical_physician

We added a bunch of testing as well (this medical_physician doesn't really have much in the way of tests)

File: medical/__openerp__.py (hunk 4 above)
Comment (pedrobaeza) [inline medical/__openerp__.py]: It's not a downgrading, because there's no internal versioning. Odoo only considers full version (for migration scripts, display...). The short version number is only a technical trick for not writing all, but we at OCA also remove that possibility by convention. The purpose of reset that part is to not have a module with version 12.0.24.1.0 in short.

Comment (lasley) [inline medical/__openerp__.py]: Awesome, makes perfect sense. Thanks for the explanation!

**Comment by lasley** [comment]
Hi all - I added a rough dependency chain in the wiki to help with reviewing of these PRs. Wasn't sure the best approach to take, so I went with text based- https://github.com/OCA/vertical-medical/wiki#dependency-chain

**Comment by lasley** [comment]
Did anyone else have any suggestions on this one?

@pedrobaeza - We have #86 open for all of the medical_physician stuff, if we can hold off on those files until then. I could remove the half-upgraded versions from this PR if you would prefer?

File: medical/models/medical_patient.py (hunk 5 above)
Comment (lasley) [inline medical/models/medical_patient.py]: @pedrobaeza Quick question if you don't mind - Is the `_` method also implicit when raising exceptions, or do we need to call it there? e.g. `raise ValidationError(_("My Error"))`

**Comment by pedrobaeza** [comment]
@lasley, for putting a translatable text raising warnings you have to use `_()`, because there you are inside Python coding. For fields and so on, it's not needed because the ORM handles them automatically.

**Comment by lasley** [comment]
Awesome, thanks for providing that distinction @pedrobaeza !

**Comment by lasley** [comment]
Alright at long last & about 5 rounds of QAs later, we have ourselves a solid product!

I am tagging all existing 9.0 PRs other than this as WIP, and have pushed current related medical commits to this one. If someone will please tag this PR with a Ready to Review, we can start the process rolling.

Once this is reviewed and merged, I will bring in the next dependency from one of the WIP PRs (which are essentially all done). This is the best process I could think of to get these merged, without adding a bunch of rebase overhead on my end. Someone please let me know if you have a better idea!

And just for reference, our current dev head is https://github.com/laslabs/vertical-medical/tree/bugfix/9.0/SD-1092-rxngo-qa - we are in the process of adding website & connector modules, so pretty much everything pre-existing & within our internal scopes is set in stone at this point.

**Comment by lasley** [comment]
This has been rebased to include all changes in our v9 medical & is now ready for review

File: medical/medical_constants.py [hunk 9]
```diff
@@ -1,24 +1,7 @@
...
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# Copyright 2004 Tech-Receptives
+# Copyright 2016 LasLabs Inc.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
 months = {
     1: "January", 2: "February", 3: "March", 4: "April",
```
Comment (lasley) [inline medical/medical_constants.py]: Please add translations to these strings

File: medical/models/medical_patient.py [hunk 10]
```diff
@@ -1,105 +1,124 @@
...
+        string='Related Partner',
+        comodel_name='res.partner',
+        required=True,
+        ondelete='cascade',
+        index=True,
+    )
+    gender = fields.Selection([
+        ('m', 'Male'),
+        ('f', 'Female'),
+    ], )
```
Comment (lasley) [inline medical/models/medical_patient.py]: Putting the `)` on the same line as the `]` defeats the purpose of the trailing `,`, which is to allow for addition of attributes without changing unrelated lines of code. Should be:

```
fields.Selection([
        ...
    ],
)
```

File: medical/models/medical_patient.py [hunk 11]
```diff
@@ -1,105 +1,124 @@
...
-            ('z', 'law marriage'),
-        ])
+    marital_status = fields.Selection([
+        ('s', 'Single'),
+        ('m', 'Married'),
+        ('w', 'Widowed'),
+        ('d', 'Divorced'),
+        ('x', 'Separated'),
+        ('z', 'law marriage'),
+    ], )
```
Comment (lasley) [inline medical/models/medical_patient.py]: Putting the `)` on the same line as the `]` defeats the purpose of the trailing `,`, which is to allow for addition of attributes without changing unrelated lines of code. Should be:

```
fields.Selection([
        ...
    ],
)
```

File: medical/models/medical_patient.py [hunk 12]
```diff
@@ -1,105 +1,124 @@
...
+                else:
+                    delta = relativedelta(now, dob)
+                    deceased = ''
+                years_months_days = '%s%s %s%s %s%s%s' % (
+                    delta.years, _('y'),
+                    delta.months, _('m'),
+                    delta.days, _('d'), deceased
+                )
+            else:
+                years_months_days = _('No DoB !')
```
Comment (lasley) [inline medical/models/medical_patient.py]: Remove the space between DoB and !

File: medical/models/medical_patient.py [hunk 13]
```diff
@@ -1,105 +1,124 @@
...
+        for rec_id in self:
+            rec_id.active = False
+            rec_id.partner_id.active = False
+
+    @api.multi
+    @api.depends('dod')
+    def _compute_deceased(self):
+        for rec_id in self:
+            if rec_id.dod:
+                rec_id.deceased = True
```
Comment (lasley) [inline medical/models/medical_patient.py]: `rec_id.deceased = bool(rec_id.dod)`

File: medical/models/res_partner.py [hunk 14]
```diff
@@ -1,37 +1,21 @@
...
+
     relationship = fields.Char(
-        'Relationship',
+        string='Relationship',
         size=25,
     )
     is_institution = fields.Boolean(
         string='Institution',
-        help='Check if the party is a Medical Center',
+        help='Check if the party is a medical center',
```
Comment (lasley) [inline medical/models/res_partner.py]: s/party/partner

File: medical/models/res_partner.py [hunk 15]
```diff
@@ -48,26 +32,31 @@ class ResPartner(models.Model):
     alias = fields.Char(
         string='Alias',
         size=256,
-        help='Common name that the Party is reffered',
+        help='Common name the party is referred to as',
```
Comment (lasley) [inline medical/models/res_partner.py]: s/party/partner

File: medical/models/res_partner.py [hunk 16]
```diff
@@ -48,26 +32,31 @@ class ResPartner(models.Model):
...
     )
     activation_date = fields.Date(
-        string='Activation date',
-        help='Date of activation of the party',
+        string='Activation Date',
+        help='Date the party was activated',
```
Comment (lasley) [inline medical/models/res_partner.py]: s/party/partner

File: medical/models/res_partner.py [hunk 17]
```diff
@@ -48,26 +32,31 @@ class ResPartner(models.Model):
...
     )
     last_name = fields.Char(
         string='Last Name',
         size=256,
-        help='Last Name',
+        help='Last name of the party',
```
Comment (lasley) [inline medical/models/res_partner.py]: s/party/partner

File: medical/models/res_partner.py [hunk 18]
```diff
@@ -48,26 +32,31 @@ class ResPartner(models.Model):
...
-        help='Common name that the Party is reffered',
+        help='Common name the party is referred to as',
     )
     activation_date = fields.Date(
-        string='Activation date',
-        help='Date of activation of the party',
+        string='Activation Date',
+        help='Date the party was activated',
     )
     last_name = fields.Char(
```
Comment (lasley) [inline medical/models/res_partner.py]: We should remove this - it's not used anywhere & will create confusion.

Please add isolation of first and last names into the roadmap as well. It should probably be in patient, not partner too.

File: medical/models/res_partner.py [hunk 19]
```diff
@@ -48,26 +32,31 @@ class ResPartner(models.Model):
...
         string='Last Name',
         size=256,
-        help='Last Name',
+        help='Last name of the party',
+    )
+    is_work = fields.Boolean(
+        string='Work',
+        help='Check if the party is a place of work',
```
Comment (lasley) [inline medical/models/res_partner.py]: s/party/partner

File: medical/models/res_partner.py [hunk 20]
```diff
@@ -48,26 +32,31 @@ class ResPartner(models.Model):
...
     )
-    is_work = fields.Boolean(string='Work')
     is_person = fields.Boolean(
         string='Person',
-        help='Check if the party is a person.',
+        help='Check if the party is a person',
```
Comment (lasley) [inline medical/models/res_partner.py]: s/party/partner

File: medical/models/res_partner.py [hunk 21]
```diff
@@ -48,26 +32,31 @@ class ResPartner(models.Model):
...
+    )
+    is_school = fields.Boolean(
```
Comment (lasley) [inline medical/models/res_partner.py:54]: Add into roadmap (for v10):

Remove `is_*` attributes on in favor of `res.partner.type` model & a M2M relation

**Comment by lasley** [comment]
@bmw95 - comments inline. Side note - Those tests look so sexy now that you're using the demo data 🎉

File: medical/medical_constants.py (hunk 9 above)
Comment (woodbrettm) [reply medical/medical_constants.py]: Fixed. Added translations to months of the year and days of the week.

Comment (pedrobaeza) [reply medical/medical_constants.py]: Isn't this month list in any place across Odoo?

File: medical/models/medical_patient.py (hunk 10 above)
Comment (woodbrettm) [reply medical/models/medical_patient.py]: Done

File: medical/models/medical_patient.py (hunk 11 above)
Comment (woodbrettm) [reply medical/models/medical_patient.py]: Done

File: medical/models/medical_patient.py (hunk 12 above)
Comment (woodbrettm) [reply medical/models/medical_patient.py]: Done. Tests updated for that.

File: medical/models/medical_patient.py (hunk 13 above)
Comment (woodbrettm) [reply medical/models/medical_patient.py]: aw nice man. Knew I was doing it a bit clunky, forget sometimes about that approach. Done.

File: medical/models/res_partner.py (hunk 14 above)
Comment (woodbrettm) [reply medical/models/res_partner.py]: Done

File: medical/models/res_partner.py (hunk 15 above)
Comment (woodbrettm) [reply medical/models/res_partner.py]: Done

File: medical/models/res_partner.py (hunk 16 above)
Comment (woodbrettm) [reply medical/models/res_partner.py]: Done

File: medical/models/res_partner.py (hunk 18 above)
Comment (woodbrettm) [reply medical/models/res_partner.py]: Removed. Added "Separation of patient/res_partner name into first and last names." to roadmap in readme.

File: medical/models/res_partner.py (hunk 19 above)
Comment (woodbrettm) [reply medical/models/res_partner.py]: Done

File: medical/models/res_partner.py (hunk 20 above)
Comment (woodbrettm) [reply medical/models/res_partner.py]: Done

File: medical/models/res_partner.py (hunk 17 above)
Comment (woodbrettm) [reply medical/models/res_partner.py]: Field removed as per comment above.

File: medical/medical_constants.py (hunk 9 above)
Comment (lasley) [reply medical/medical_constants.py]: I have no idea TBH. This is leftover from legacy & I'm not even 100% where it's all being used. We'll probably remove entirely in v10, I just haven't had time to audit fully

File: medical/models/res_partner.py (hunk 20 above)
Comment (lasley) [reply medical/models/res_partner.py]: No need to note done on each of the items. You only need comment on them if you have a question, comment, etc. When fully done, just drop a global comment advising of such

File: medical/models/res_partner.py (hunk 21 above)
Comment (woodbrettm) [reply medical/models/res_partner.py:54]: - Added: "(Odoo v10) Remove all is_\* fields (e.g. is_school, is_work, etc.) from res.partner and move to res.partner.type."
- Visually grouped all the is_\* fields together in the res.partner.py file

File: medical/medical_constants.py (hunk 9 above)
Comment (pedrobaeza) [reply medical/medical_constants.py]: OK, let's move this way for now.

Comment (lasley) [reply medical/medical_constants.py]: @bmw95 - Please add a note in the ReadMe roadmap regarding the removal of medical constants

Comment (woodbrettm) [reply medical/medical_constants.py]: - Is [this](https://github.com/odoo/odoo/blob/9c8d8b90c0f09575722548881fe812faf9547ffe/openerp/models.py#L1927) something similar? After some searching I noticed the time_intervals dictionary (on line 1943) used in several base_ addons.
- Couple addons I noticed using the months etc from medical (medical_physician being one).
- **Do you guys want me to add this to the roadmap section in the readme?**

Comment (woodbrettm) [reply medical/medical_constants.py]: oh my bad sorry @lasley didn't catch your comment in time

Comment (lasley) [reply medical/medical_constants.py]: Nope `time_intervals` are deltas for using w/ math. These are meant for select fields (I think).

File: medical/models/medical_patient.py (hunk 10 above)
Comment (woodbrettm) [reply medical/models/medical_patient.py]: @lasley Getting lint errors from that way. Can I do:

```
fields.Selection(
    [
        ...
    ],
)
```

?

**Comment by woodbrettm** [comment]
Ready for further review :)

File: medical/models/res_partner.py [hunk 22]
```diff
@@ -13,62 +13,57 @@ class ResPartner(models.Model):
...
     )
     is_work = fields.Boolean(
         string='Work',
-        help='Check if the party is a place of work',
+        help='Check if the s/party/partner is a place of work',
     )
     is_person = fields.Boolean(
         string='Person',
-        help='Check if the party is a person',
+        help='Check if the s/party/partner is a person',
```
Comment (lasley) [inline medical/models/res_partner.py]: Lmao `s/party/partner` means spelling error. Change word(s) on left side with words on right. This should say `partner`

Comment (woodbrettm) [reply medical/models/res_partner.py]: oh hahahaha. Yeah wasn't quite sure what that meant.

Comment (lasley) [reply medical/models/res_partner.py]: Ask in chat next time 😉

**Comment by lasley** [comment]
Thanks @bmw95 - LGTM 👍

**Comment by lasley** [comment]
@t3ddftw please review when you get a chance.

File: medical/models/res_partner.py [hunk 23]
```diff
@@ -1,85 +1,69 @@
...
 
 from openerp import fields, models
 
 
 class ResPartner(models.Model):
     _inherit = 'res.partner'
+
     relationship = fields.Char(
-        'Relationship',
+        string='Relationship',
```
Comment (tedsalmon) [inline medical/models/res_partner.py]: This is superfluous given the field name.

**Comment by tedsalmon** [comment]
LGTM aside from the minor detail I noted. 👍
